leku/src/main/java/com/schibstedspain/leku/search/LekuSearchFragment.java

```diff
@@ -0,0 +1,252 @@
+    toolbar.inflateMenu(R.menu.toolbar_menu);
+    searchOption = toolbar.getMenu().findItem(R.id.action_voice);
+    toolbar.setOnMenuItemClickListener(this::onToolbarMenuItemSelected);
+  }
+
+  private boolean onToolbarMenuItemSelected(MenuItem item) {
+    if (item.getItemId() == R.id.action_voice) {
+      onVoiceItemSelected();
+      return true;
+    }
+    return false;
+  }
+
+  public void clear() {
+    searchView.setText("");
+    closeKeyboard();
+  }
+
+  private void onVoiceItemSelected() {
+    if (searchView.getText().toString().isEmpty()) {
+      startVoiceRecognitionActivity();
+    } else {
+      retrieveLocationFrom(searchView.getText().toString());
+      closeKeyboard();
+    }
+  }
+
+  private void startVoiceRecognitionActivity() {
+    if (checkPlayServices()) {
+      Intent intent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
+      intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM);
+      intent.putExtra(RecognizerIntent.EXTRA_PROMPT, getString(R.string.voice_search_promp));
+      intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE, getString(R.string.voice_search_extra_language));
+
+      try {
+        startActivityForResult(intent, REQUEST_PLACE_PICKER);
+      } catch (ActivityNotFoundException e) {
+        track(TrackEvents.noVoiceRecognition);
+      }
+    }
+  }
+
+  protected void track(TrackEvents event) {
+    LocationPicker.getTracker().onEventTracked(event);
+  }
+
+  @Override
+  public void onAttach(Context context) {
+    super.onAttach(context);
+
+    if (context instanceof LekuSearchCallback) {
+      lekuSearchCallback = (LekuSearchCallback) context;
+    } else {
+      lekuSearchCallback = NullView.createFor(LekuSearchCallback.class);
+    }
+  }
+
+  @Override
+  public void onDetach() {
+    lekuSearchCallback = NullView.createFor(LekuSearchCallback.class);
+    super.onDetach();
+  }
+
+  private boolean checkPlayServices() {
+    GoogleApiAvailability googleAPI = GoogleApiAvailability.getInstance();
+    int result = googleAPI.isGooglePlayServicesAvailable(getContext());
+    if (result != ConnectionResult.SUCCESS) {
+      if (googleAPI.isUserResolvableError(result)) {
+        googleAPI.getErrorDialog(getActivity(), result, CONNECTION_FAILURE_RESOLUTION_REQUEST).show();
+      }
+      return false;
+    }
+    return true;
+  }
+
+  private void setUpClearListener() {
+    clearSearchButton.setOnClickListener(view -> searchView.setText(""));
+  }
+
+  private void setUpSearchViewListeners() {
+    searchView.setOnEditorActionListener((v, actionId, event) -> {
+      boolean handled = false;
+      if (actionId == EditorInfo.IME_ACTION_SEARCH) {
+        retrieveLocationFrom(v.getText().toString());
+        closeKeyboard();
+        handled = true;
+      }
+      return handled;
+    });
+    textWatcher = getSearchTextWatcher();
+    searchView.addTextChangedListener(textWatcher);
+  }
+
+  private void closeKeyboard() {
```

Comment:
We could move this method outside this class. It's the common `SoftkeyboardManager` or something like that